crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -279,66 +295,343 @@ where
+            // gen_ibc_shielded_transfer in the non-refund case
+            let ibc_denom = self.query_ibc_denom(
+                packet_data.token.denom.to_string(),
+                Some(&Address::Internal(InternalAddress::Ibc)),
+            )?;
+            let token = namada_ibc::received_ibc_token(
+                ibc_denom,
+                &msg.message.packet.port_id_on_a,
+                &msg.message.packet.chan_id_on_a,
+                &msg.message.packet.port_id_on_b,
+                &msg.message.packet.chan_id_on_b,
+            )
+            .into_storage_result()
+            .map_err(Error::NativeVpError)?;
+            let delta = ValueSum::from_pair(
+                token.clone(),
+                Amount::from_uint(Uint(*packet_data.token.amount), 0).unwrap(),
+            );
+            // Enable funds to be taken from the IBC internal
+            // address and be deposited elsewhere
+            // Required for the IBC internal Address to release
+            // funds
+            let pre_entry =
+                acc.pre.entry(addr_taddr(IBC)).or_insert(ValueSum::zero());
+            *pre_entry =
+                checked!(pre_entry + &delta).map_err(native_vp::Error::new)?;
+        }
+        Ok(())
+    }
+
+    // Apply relevant IBC packets to the changed balances structure
+    fn apply_ibc_packet(
+        &self,
+        mut acc: ChangedBalances,
+        ibc_msg: &IbcMessage,
+        keys_changed: &BTreeSet<Key>,
+        ibc_acks: &BTreeMap<Vec<u8>, AcknowledgementStatus>,
+    ) -> Result<ChangedBalances> {
+        match ibc_msg {
+            // This event is emitted on the sender
+            IbcMessage::Transfer(msg) => {
+                // Get the packet commitment from post-storage that corresponds
+                // to this event
+                let receiver = msg.message.packet_data.receiver.to_string();
+                let addr = TAddrData::Ibc(receiver.clone());
+                acc.decoder.insert(ibc_taddr(receiver), addr);
+                self.apply_send_packet(&mut acc, msg, keys_changed)?;
+            }
+            // This event is emitted on the receiver
+            IbcMessage::RecvPacket(msg) => {
+                // Check if this is a Transfer packet
+                let Ok(packet_data) = serde_json::from_slice::<PacketData>(
+                    &msg.message.packet.data,
+                ) else {
+                    return Ok(acc);
+                };
+                let receiver = packet_data.receiver.to_string();
+                let addr = TAddrData::Ibc(receiver.clone());
+                acc.decoder.insert(ibc_taddr(receiver), addr);
+                self.apply_write_ack(
+                    &mut acc,
+                    msg,
+                    &packet_data,
+                    keys_changed,
+                    ibc_acks,
+                )?;
+            }
+            // Ignore all other IBC events
+            _ => {}
```

Comment:
IBC messages `AckPacket` and `Timeout` could bring MASP transfer due to the refund when IBC transfer failure too.

Reply:
I believe that the changes caused by an `AckPacket` message are already captured by `apply_balance_change` . For example see the following excerpt generated by running `run_ledger_ibc_with_hermes`:
```
IBC Message: AckPacket(MsgAcknowledgement { message: MsgAcknowledgement { packet: Packet { sequence: Sequence(3), source_port: PortId(\"transfer\"), source_channel: ChannelId(\"channel-0\"), destination_port: PortId(\"transfer\"), destination_channel: ChannelId(\"channel-0\"), data: [123, 34, 100, 101, 110, 111, 109, 34, 58, 34, 116, 110, 97, 109, 49, 113, 121, 56, 113, 103, 120, 108, 99, 116, 101, 101, 104, 108, 107, 55, 48, 115, 110, 56, 119, 120, 50, 112, 100, 108, 97, 118, 116, 97, 121, 112, 51, 56, 118, 118, 114, 110, 107, 104, 113, 34, 44, 34, 97, 109, 111, 117, 110, 116, 34, 58, 34, 49, 48, 48, 48, 48, 48, 48, 48, 48, 48, 34, 44, 34, 115, 101, 110, 100, 101, 114, 34, 58, 34, 122, 110, 97, 109, 49, 99, 104, 117, 99, 109, 110, 115, 118, 107, 117, 109, 121, 52, 52, 110, 48, 120, 113, 54, 112, 116, 102, 110, 108, 48, 112, 54, 117, 51, 52, 56, 102, 55, 109, 117, 120, 97, 50, 113, 55, 116, 108, 103, 114, 53, 118, 56, 97, 116, 54, 113, 118, 108, 57, 112, 121, 52, 97, 112, 101, 117, 99, 121, 118, 112, 55, 97, 120, 55, 122, 122, 53, 122, 99, 114, 34, 44, 34, 114, 101, 99, 101, 105, 118, 101, 114, 34, 58, 34, 105, 110, 118, 97, 108, 105, 100, 95, 114, 101, 99, 101, 105, 118, 101, 114, 34, 44, 34, 109, 101, 109, 111, 34, 58, 34, 34, 125], timeout_height: Never, timeout_timestamp: Timestamp { time: Some(Time(2024-06-14 13:52:21.516949312)) } }, acknowledgement: Acknowledgement([123, 34, 101, 114, 114, 111, 114, 34, 58, 34, 102, 97, 105, 108, 101, 100, 32, 116, 111, 32, 112, 97, 114, 115, 101, 32, 97, 115, 32, 65, 99, 99, 111, 117, 110, 116, 73, 100, 34, 125]), proof_acked_on_b: CommitmentProof(\"0A97020A94020A3261636B732F706F7274732F7472616E736665722F6368616E6E656C732F6368616E6E656C2D302F73657175656E6365732F3312200402EE754C5E43142130AC8C370D449337CDF35722F4C1A36F9AF37B10A687B31A2408012A200000000000000000000000000000000000000000000000000000000000000000222408011220E062C9E219D7FBD9D42B8871745637580097DF41B9BBE2C2629884CD7CF5C428222408011A20007159D6694F92C048952FABE29A83CC39BC9526AAD27F0FBB3E142E9C54C854222408011A201D27263E79026BA20EC88F32874CD5589123B3E2B0468657E94B9B07C96C772D222408011A20F3017AA4EF352D4585DD8E16C56F868C71269B3DBAD67873BF4BD38AC8592C250A9E010A9B010A036962631220394803D0AF9DBCD091308DC5BE9DE0BBA22A068EB0CEC05A4134A8ADFD9EA30A1A26080110012A200000000000000000000000000000000000000000000000000000000000000000222408011A203D02BB9C9E71792AEFC7924B284E76883A7043CB5768CE7E33717CCBFA5A7BD8222408011A201E79C6A260250336321A4347CAA7F354672D7FA9839A02CA58762EFE0A9DBB7D\"), proof_height_on_b: Height { revision: 0, height: 124 }, signer: Signer(\"tnam1qz0nvec686e9pks8ynhm5ddq8ke7j2eey50uagtr\") }, transfer: Some(ShieldingTransfer { source: Internal IBC: tnam1qcqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqvtr7x4, token: Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq, amount: DenominatedAmount { amount: Amount { raw: 1000000000 }, denom: Denomination(0) }, shielded_section_hash: Hash([65, 170, 255, 182, 66, 245, 59, 220, 144, 51, 109, 85, 74, 210, 181, 200, 182, 203, 57, 224, 222, 167, 45, 26, 94, 17, 60, 80, 101, 34, 103, 187]) }) })
Changed Balances: ChangedBalances { tokens: {AssetType { identifier: [10, 170, 166, 12, 65, 99, 62, 13, 116, 182, 9, 215, 18, 196, 130, 27, 41, 214, 181, 104, 168, 46, 202, 240, 12, 92, 76, 92, 181, 215, 209, 118], nonce: Some(0) }: (Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq, Denomination(8), Three), AssetType { identifier: [53, 212, 33, 11, 138, 34, 147, 17, 161, 183, 27, 205, 164, 179, 27, 28, 156, 104, 183, 65, 47, 108, 70, 245, 193, 214, 195, 181, 180, 141, 165, 106], nonce: Some(2) }: (Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq, Denomination(8), One), AssetType { identifier: [96, 72, 232, 131, 106, 52, 7, 194, 177, 23, 31, 184, 184, 50, 168, 110, 65, 206, 104, 151, 235, 80, 90, 101, 169, 126, 226, 45, 185, 89, 59, 212], nonce: Some(1) }: (Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq, Denomination(8), Two), AssetType { identifier: [161, 97, 138, 24, 204, 180, 226, 116, 36, 16, 242, 35, 132, 48, 211, 54, 131, 69, 170, 62, 15, 49, 26, 131, 92, 95, 187, 63, 135, 128, 197, 147], nonce: Some(1) }: (Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq, Denomination(8), Zero)}, decoder: {TransparentAddress([155, 209, 98, 252, 116, 194, 185, 61, 115, 5, 201, 229, 4, 170, 177, 214, 152, 246, 217, 52]): Addr(Internal IBC: tnam1qcqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqvtr7x4), TransparentAddress([196, 128, 240, 153, 79, 139, 252, 48, 1, 179, 248, 81, 97, 118, 59, 124, 233, 55, 127, 247]): Addr(Internal MASP: tnam1pcqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqzmefah)}, ibc_denoms: {\"tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq\": Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq}, pre: {TransparentAddress([155, 209, 98, 252, 116, 194, 185, 61, 115, 5, 201, 229, 4, 170, 177, 214, 152, 246, 217, 52]): ValueSum({Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq: Amount { raw: 2000000000 }}), TransparentAddress([196, 128, 240, 153, 79, 139, 252, 48, 1, 179, 248, 81, 97, 118, 59, 124, 233, 55, 127, 247]): ValueSum({Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq: Amount { raw: 8000000000 }})}, post: {TransparentAddress([155, 209, 98, 252, 116, 194, 185, 61, 115, 5, 201, 229, 4, 170, 177, 214, 152, 246, 217, 52]): ValueSum({Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq: Amount { raw: 1000000000 }}), TransparentAddress([196, 128, 240, 153, 79, 139, 252, 48, 1, 179, 248, 81, 97, 118, 59, 124, 233, 55, 127, 247]): ValueSum({Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq: Amount { raw: 9000000000 }})} }
```
At a high level this packet is refunding 10 `tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq` tokens. Note how the IBC internal address balance goes down by 10 tokens and the MASP internal address balance goes up by 10 tokens. This suggests to me that it is not necessary to utilize the `AckPacket` messages here, and that monitoring the balance changes of IBC internal address will suffice.

Reply:
I also believe that an argument analogous to the above can be made for the IBC `Timeout` message. For example, consider the following excerpt from the same test:
```
IBC Message: Timeout(MsgTimeout { message: MsgTimeout { packet: Packet { sequence: Sequence(4), source_port: PortId(\"transfer\"), source_channel: ChannelId(\"channel-0\"), destination_port: PortId(\"transfer\"), destination_channel: ChannelId(\"channel-0\"), data: [123, 34, 100, 101, 110, 111, 109, 34, 58, 34, 116, 110, 97, 109, 49, 113, 121, 56, 113, 103, 120, 108, 99, 116, 101, 101, 104, 108, 107, 55, 48, 115, 110, 56, 119, 120, 50, 112, 100, 108, 97, 118, 116, 97, 121, 112, 51, 56, 118, 118, 114, 110, 107, 104, 113, 34, 44, 34, 97, 109, 111, 117, 110, 116, 34, 58, 34, 49, 48, 48, 48, 48, 48, 48, 48, 48, 48, 34, 44, 34, 115, 101, 110, 100, 101, 114, 34, 58, 34, 122, 110, 97, 109, 49, 118, 112, 50, 102, 104, 121, 57, 101, 56, 100, 119, 56, 53, 51, 122, 99, 97, 50, 116, 55, 112, 120, 120, 53, 120, 52, 113, 115, 108, 50, 121, 118, 103, 114, 109, 97, 107, 55, 101, 114, 48, 56, 57, 53, 118, 115, 57, 51, 114, 115, 55, 101, 118, 102, 114, 112, 122, 116, 97, 109, 108, 122, 51, 115, 109, 121, 53, 57, 54, 104, 103, 115, 100, 97, 112, 34, 44, 34, 114, 101, 99, 101, 105, 118, 101, 114, 34, 58, 34, 122, 110, 97, 109, 49, 122, 120, 116, 56, 101, 50, 50, 117, 122, 54, 54, 54, 99, 101, 55, 104, 120, 113, 112, 99, 54, 57, 121, 102, 106, 51, 116, 112, 100, 57, 118, 50, 54, 101, 112, 50, 101, 112, 119, 110, 51, 52, 107, 118, 121, 117, 119, 106, 104, 57, 56, 104, 104, 114, 101, 57, 56, 57, 55, 115, 104, 99, 106, 106, 52, 99, 110, 113, 117, 103, 119, 108, 118, 52, 113, 34, 44, 34, 109, 101, 109, 111, 34, 58, 34, 34, 125], timeout_height: Never, timeout_timestamp: Timestamp { time: Some(Time(2024-06-14 12:52:47.695933275)) } }, next_seq_recv_on_b: Sequence(4), proof_unreceived_on_b: CommitmentProof(\"0A86041283040A3672656365697074732F706F7274732F7472616E736665722F6368616E6E656C732F6368616E6E656C2D302F73657175656E6365732F3412F9010A3672656365697074732F706F7274732F7472616E736665722F6368616E6E656C732F6368616E6E656C2D302F73657175656E6365732F331201011A2408012A200000000000000000000000000000000000000000000000000000000000000000222408011220FE65AAC7FA573BCAE82431882B3FB689926FC708AA848F0C15852A2A17152625222408011220FB867EFBA26D5059137DFA54C2A8DC00673F55A015E41D9078A8875B571B6BA0222408011A2071376AE7874E57AFDBC1088887A3DB66277BEFF0C3170A79E4C655F63602EFA4222408011220C9A504EFDEDC6E7BA6C7534AE6B493732C04C356E8167A12EFE2BE43EC51F0ED1ACC010A3677697468647261772F746E616D31703575387A79766D667633726C32376D6333677367617870613233396875656839736774637A3874122000743BA40B0000000000000000000000000000000000000000000000000000001A2408012A20000000000000000000000000000000000000000000000000000000000000000022240801122076D3570FAFF45DCDE578C121C9EE89FCBBA38E469D6F2D1F904E4198FF263767222408011220C9A504EFDEDC6E7BA6C7534AE6B493732C04C356E8167A12EFE2BE43EC51F0ED0A9E010A9B010A036962631220394803D0AF9DBCD091308DC5BE9DE0BBA22A068EB0CEC05A4134A8ADFD9EA30A1A26080110012A200000000000000000000000000000000000000000000000000000000000000000222408011A2014AD5A9623535DFFB0A8F1691AB21B9D133D6AF9E56A66AD6C0F93CBE4132190222408011A201E79C6A260250336321A4347CAA7F354672D7FA9839A02CA58762EFE0A9DBB7D\"), proof_height_on_b: Height { revision: 0, height: 150 }, signer: Signer(\"tnam1qz0nvec686e9pks8ynhm5ddq8ke7j2eey50uagtr\") }, transfer: Some(ShieldingTransfer { source: Internal IBC: tnam1qcqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqvtr7x4, token: Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq, amount: DenominatedAmount { amount: Amount { raw: 1000000000 }, denom: Denomination(0) }, shielded_section_hash: Hash([214, 51, 205, 51, 14, 5, 69, 21, 36, 81, 90, 101, 230, 220, 29, 243, 81, 31, 203, 134, 134, 120, 15, 247, 71, 210, 170, 44, 7, 40, 14, 60]) }) })
Changed Balances: ChangedBalances { tokens: {AssetType { identifier: [10, 170, 166, 12, 65, 99, 62, 13, 116, 182, 9, 215, 18, 196, 130, 27, 41, 214, 181, 104, 168, 46, 202, 240, 12, 92, 76, 92, 181, 215, 209, 118], nonce: Some(0) }: (Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq, Denomination(8), Three), AssetType { identifier: [53, 212, 33, 11, 138, 34, 147, 17, 161, 183, 27, 205, 164, 179, 27, 28, 156, 104, 183, 65, 47, 108, 70, 245, 193, 214, 195, 181, 180, 141, 165, 106], nonce: Some(2) }: (Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq, Denomination(8), One), AssetType { identifier: [96, 72, 232, 131, 106, 52, 7, 194, 177, 23, 31, 184, 184, 50, 168, 110, 65, 206, 104, 151, 235, 80, 90, 101, 169, 126, 226, 45, 185, 89, 59, 212], nonce: Some(1) }: (Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq, Denomination(8), Two), AssetType { identifier: [161, 97, 138, 24, 204, 180, 226, 116, 36, 16, 242, 35, 132, 48, 211, 54, 131, 69, 170, 62, 15, 49, 26, 131, 92, 95, 187, 63, 135, 128, 197, 147], nonce: Some(1) }: (Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq, Denomination(8), Zero)}, decoder: {TransparentAddress([155, 209, 98, 252, 116, 194, 185, 61, 115, 5, 201, 229, 4, 170, 177, 214, 152, 246, 217, 52]): Addr(Internal IBC: tnam1qcqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqvtr7x4), TransparentAddress([196, 128, 240, 153, 79, 139, 252, 48, 1, 179, 248, 81, 97, 118, 59, 124, 233, 55, 127, 247]): Addr(Internal MASP: tnam1pcqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqzmefah)}, ibc_denoms: {\"tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq\": Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq}, pre: {TransparentAddress([155, 209, 98, 252, 116, 194, 185, 61, 115, 5, 201, 229, 4, 170, 177, 214, 152, 246, 217, 52]): ValueSum({Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq: Amount { raw: 2000000000 }}), TransparentAddress([196, 128, 240, 153, 79, 139, 252, 48, 1, 179, 248, 81, 97, 118, 59, 124, 233, 55, 127, 247]): ValueSum({Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq: Amount { raw: 8000000000 }})}, post: {TransparentAddress([155, 209, 98, 252, 116, 194, 185, 61, 115, 5, 201, 229, 4, 170, 177, 214, 152, 246, 217, 52]): ValueSum({Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq: Amount { raw: 1000000000 }}), TransparentAddress([196, 128, 240, 153, 79, 139, 252, 48, 1, 179, 248, 81, 97, 118, 59, 124, 233, 55, 127, 247]): ValueSum({Established: tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq: Amount { raw: 9000000000 }})} }
```
This packet is also refunding 10 `tnam1qy8qgxlcteehlk70sn8wx2pdlavtayp38vvrnkhq` tokens. And note how the `Timeout` message was not utilized.